ui/panels-plugin/src/plugins/scatterplot/ScatterChartPanel.tsx

```diff
@@ -0,0 +1,165 @@
+  startTime: Date;
+}
+
+export function getUnixTimeRange(timeRange: AbsoluteTimeRange) {
+  const { start, end } = timeRange;
+  return {
+    start: Math.ceil(getUnixTime(start)),
+    end: Math.ceil(getUnixTime(end)),
+  };
+}
+
+const generateErrorAlert = (message: string) => {
+  const alertMessage = {
+    name: message,
+    message: message,
+  };
+  return (
+    <div data-testid="ScatterChartPanel_ErrorAlert">
+      <ErrorAlert error={alertMessage} />
+    </div>
+  );
+};
+
+export type ScatterChartPanelProps = PanelProps<ScatterChartOptions>;
+
+/**
+ * ScatterChartPanel receives data from the DataQueriesProvider and transforms it
+ * into a `dataset` object that Apache ECharts can consume. Additionally,
+ * data formatting is also dictated in this component. Formatting includes
+ * datapoint size and color.
+ *
+ * Documentation for data structures accepted by Apache ECharts:
+ *  https://echarts.apache.org/handbook/en/concepts/dataset
+ *
+ * Examples for scatter chart formatting in Apache ECharts:
+ *  https://echarts.apache.org/examples/en/index.html#chart-type-scatter
+ *
+ * @returns a `ScatterPlot` component that contains an EChart which will handle
+ * visuzliation of the data.
+ */
+export function ScatterChartPanel(props: ScatterChartPanelProps) {
+  const { contentDimensions } = props;
+  const { queryResults: traceResults, isLoading: traceIsLoading } = useDataQueries('TraceQuery');
+
+  // Generate dataset
+  // Transform Tempo API response to fit 'dataset' structure from Apache ECharts
+  // https://echarts.apache.org/handbook/en/concepts/dataset
+  const dataset = useMemo(() => {
+    const traceData = traceResults[0]?.data;
+    if (traceIsLoading || traceData === undefined) {
+      return [];
+    }
+    const dataset = [];
+    for (const result of traceResults) {
+      if (traceIsLoading || traceData === undefined) {
+        return [];
+      }
+      if (result.isLoading || result.data === undefined) continue;
+      const dataSeries = result.data.traces.map((trace) => {
+        const newTraceValue: EChartTraceValue = {
+          ...trace,
+          startTime: new Date(trace.startTimeUnixMs), // convert unix epoch time to Date
+        };
+        return newTraceValue;
+      });
+      dataset.push({
+        source: dataSeries,
+      });
+    }
+    return dataset;
+  }, [traceIsLoading, traceResults]);
+
+  // Error check: specify an alert if no traces are returned from the query
+  const traceData = traceResults[0]?.data;
+  if (!traceIsLoading && traceData?.traces.length === 0) {
+    const query = traceData?.metadata?.executedQueryString;
+    return generateErrorAlert(`No traces found for the query : " ${query} " .`);
+  }
+
+  // Formatting for the dataset
+  // 1. Map x,y coordinates
+  // 2. Datapoint size corresponds to the number of spans in a trace
+  // 3. Color datapoint red if the trace contains an error
+  const seriesTemplate: SeriesOption = {
```

Comment:
The dataset lines above are in a useMemo so that's good, but these other lines could benefit from being memozied too. It would break the rules of hooks though so having a separate component like ScatterChartPanelLoaded.tsx would be the way to go (other panel types would benefit from that too)

Reply:
TBH, I’m a novice with hooks and don’t see how it would break the rules. Could you elaborate on the rules that are being broken and the implementation you’d like to see for the separate component `ScatterChartPanelLoaded`? In the latest push, I attempted to use the useMemo() hook directly in scatterChartPanel to memoize the series formatting, and it didn’t throw any errors…

I am not sure how the [rules](https://react.dev/warnings/invalid-hook-call-warning)](https://react.dev/warnings/invalid-hook-call-warning) are being broken because the memoized function is not inside a condition, loop, or event handler, and there are no other hooks within the function. And I moved the error checks that would return the function early, after the hooks.

Reply:
Oh great! I thought it was after the loading if conditions but I might have misread too, if no errors were thrown or anything then you should be good